statsmodels/stats/_cov_tools.py

```diff
@@ -0,0 +1,463 @@
+
+    """
+    n = corr.shape[0]   # k_vars, n is used in reference
+    # Neudecker 1996 eq (4), notation Kd is Md
+
+    # Note dropping the outer Dup doesn't work
+    outer = np.eye(n**2) - (np.kron(np.eye(n), corr).dot(Md(n)))
+    cov_corr_ = outer.dot(np.kron(corr, corr)).dot(outer.T)
+    dd = Dup(n)
+    if drop:
+        dd = dd[:, dd.sum(0) == 2]
+    cov_corr_ = (dd.T.dot(cov_corr_).dot(dd)) / 2
+    if kurt != 0:
+        cov_corr_ *= (1 + kurt)
+    return cov_corr_ / nobs
+
+
+def _cov_cov_vech(cov, nobs):
+    """covariance of vech of sample covariance assuming normality
+
+    This is the inverse of the information matrix for the lower triangular
+    elements for estimating cov.
+
+    This is a function just to store a formula. It will be better to reuse
+    intermediate results.
+
+    """
+    k_vars = cov.shape[0]
+    d2 = Dup(k_vars)
+    d2_pinv = np.linalg.pinv(d2)  # this is D^+
+    # from Magnus Neudecker:
+    cc_vech = 2 * d2_pinv.dot(np.kron(cov, cov)).dot(d2_pinv.T) / nobs
+    return cc_vech
+
+
+def _gls_cov_vech(data, cov0=None, ddof=0):
+    """gls using weight matrix assuming normal distributed data
+
+    Another function just to store some pieces.
+    This needs to be extended to linear functions of vech(cov).
+    """
+    data = np.asarray(data)
+    nobs, k_vars = data.shape
+    data = data - data.mean(0)
+    if cov0 is None:
+        cov0 = data.T.dot(data) / (nobs - ddof)
+
+    weights = _cov_cov_vech(cov0, nobs)
+    # the following should impose restrictions and use residuals
+    mom = vech_cross_product(data, data)
+    momcond = mom.mean(0)   # I think this is currently be just vech(cov_sample)
+    gls_qf = momcond.dot(weights).dot(momcond)
+    return gls_qf
+
+
+def mom4(data, method='einsum', ddof=0):
+    """empirical 4th moment of data, not centered
+
+    To compute the centered moment, subtract the mean from data.
+
+    Parameters
+    ----------
+    data : array_like
+        2-D data with observations in rows and variables in columns
+    method : str
+        Computational method, mainly used as verification and for testing.
+        If method is 'einsum', then numpy.einsum is used. Otherwise, a Python
+        loop over all observations is used. Results are identical up to
+        floating point precision in the computation.
+    ddof : float
+        degrees of freedom correction in denominator. Default is ddof=0 and
+        the denominator is nobs
+
+    Returns
+    -------
+    mom4 : ndarray
+        2-D array raveled according to vec(cov).
+
+    """
+    y = np.asarray(data)
+    nobs = y.shape[0]
+    if method == 'einsum':
+        m4 = np.einsum('ti,tj,tk,th->ijkh', y, y, y, y)
+        k_vec = m4.shape[0] ** 2
+        m4 = m4.reshape((k_vec, k_vec))
+    else:
+        m4 = 0
+        for row in y:
+            c = np.outer(row, row)  # outer_product
+            m4 += np.kron(c, c)
+
+    return m4 / (nobs - ddof)
+
+
+def _mom4_normal(i, j, k, h, cov):
```

Comment:
useful mainly for cross-checking, in most cases we want the kronecker product version